File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3128_right-triangles/solutions/python3/solution.py
from typing import List

class Solution:
    def numberOfRightTriangles(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        
        # Compute counts of 1's in each row and each column
        rowCount = [sum(row) for row in grid]
        colCount = [sum(grid[i][j] for i in range(rows)) for j in range(cols)]
        
        # Count the number of right triangles
        triangles = 0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1:
                    # Only if there is a '1' at the current position
                    # (i, j) can be the right-angle vertex of a triangle
                    triangles += (rowCount[i] - 1) * (colCount[j] - 1)
        
        return triangles

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3128_right-triangles/solutions/java/solution.java
class Solution {
    public long numberOfRightTriangles(int[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        long count = 0;
        
        // Count the number of 1s in each row and each column
        int[] rowCount = new int[rows];
        int[] colCount = new int[cols];
        
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 1) {
                    rowCount[r]++;
                    colCount[c]++;
                }
            }
        }
        
        // Calculate the number of right triangles using the intersection points
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 1) {
                    // (rowCount[r] - 1) choices in the row, (colCount[c] - 1) choices in the column
                    count += (rowCount[r] - 1) * (colCount[c] - 1);
                }
            }
        }
        
        return count;
    }
}
